pkg/doc/jose/jws.go

```diff
@@ -0,0 +1,380 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package jose
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/square/go-jose/v3"
+	"github.com/square/go-jose/v3/json"
+)
+
+const (
+	jwsPartsCount    = 3
+	jwsHeaderPart    = 0
+	jwsPayloadPart   = 1
+	jwsSignaturePart = 2
+)
+
+// Headers represents JOSE headers.
+type Headers map[string]interface{}
+
+// JWK (JSON Web Key) is a JSON data structure that represents a cryptographic key.
+type JWK jose.JSONWebKey
+
+// JWK gets JWK from JOSE headers.
+func (h Headers) JWK() (*JWK, bool) {
+	jwkRaw, ok := h[HeaderJSONWebKey]
+	if !ok {
+		return nil, false
+	}
+
+	var jwk JWK
+
+	err := convertMapToValue(jwkRaw, &jwk)
+	if err != nil {
+		return nil, false
+	}
+
+	return &jwk, true
+}
+
+// KeyID gets Key ID from JOSE headers.
+func (h Headers) KeyID() (string, bool) {
+	return h.stringValue(HeaderKeyID)
+}
+
+// Algorithm gets Key ID from JOSE headers.
+func (h Headers) Algorithm() (string, bool) {
+	return h.stringValue(HeaderAlgorithm)
+}
+
+func (h Headers) stringValue(key string) (string, bool) {
+	kRaw, ok := h[key]
+	if !ok {
+		return "", false
+	}
+
+	kStr, ok := kRaw.(string)
+
+	return kStr, ok
+}
+
+// JSONWebSignature defines JSON Web Signature (https://tools.ietf.org/html/rfc7515)
+type JSONWebSignature struct {
+	ProtectedHeaders   Headers
+	UnprotectedHeaders Headers
+	Payload            []byte
+
+	signature []byte
+}
+
+// SignatureVerifier makes verification of JSON Web Signature.
+type SignatureVerifier interface {
+	// Verify verifies JWS based on the signing input.
+	Verify(joseHeaders Headers, payload, signingInput, signature []byte) error
+}
+
+// SignatureVerifierFunc is a function wrapper for SignatureVerifier.
+type SignatureVerifierFunc func(joseHeaders Headers, payload, signingInput, signature []byte) error
+
+// Verify verifies JWS signature.
+func (s SignatureVerifierFunc) Verify(joseHeaders Headers, payload, signingInput, signature []byte) error {
+	return s(joseHeaders, payload, signingInput, signature)
+}
+
+// CompositeAlgSignatureVerifier defines composite signature verifier based on the algorithm
+// taken from JOSE header alg.
+type CompositeAlgSignatureVerifier struct {
+	verifierByAlg map[string]SignatureVerifier
+}
+
+// NewCompositeAlgSignatureVerifier creates a new CompositeAlgSignatureVerifier
+func NewCompositeAlgSignatureVerifier() *CompositeAlgSignatureVerifier {
+	return &CompositeAlgSignatureVerifier{
+		verifierByAlg: make(map[string]SignatureVerifier),
+	}
+}
+
+// AddVerifier adds a new verifier of algorithm.
+// If algorithm verifier is already present, it's get rewritten.
+func (v *CompositeAlgSignatureVerifier) AddVerifier(
```

Comment:
I assume we expect to add to this structure over time instead of just constructing with Verifiers?

Reply:
I agree, no sense to make it mutable. I removed `AddVerifier()` method.